crates/bevy_ecs/src/change_detection.rs

```diff
@@ -271,6 +283,143 @@ change_detection_impl!(Mut<'a, T>, T,);
+/// a batch changes in unison.  A batch has changed if any of its elements have changed.
+pub struct MutBatch<'a, T: ?Sized, const N: usize, const ALIGN: usize>
+where
+    ComponentTicks: AlignedBatchGat<N, ALIGN>,
+    T: AlignedBatchGat<N, ALIGN>,
+{
+    pub(crate) value: &'a mut AlignedBatchT<T, N, ALIGN>,
+    pub(crate) ticks: TicksBatch<'a, N, ALIGN>,
+    pub(crate) _marker: PhantomData<T>,
+}
+
+impl<'a, T: ?Sized, const N: usize, const ALIGN: usize> DetectChanges for MutBatch<'a, T, N, ALIGN>
+where
+    ComponentTicks: AlignedBatchGat<N, ALIGN>,
+    T: AlignedBatchGat<N, ALIGN>,
+{
+    #[inline]
+    fn is_added(&self) -> bool {
+        self.ticks
+            .component_ticks
+            .as_array()
+            .iter()
+            .any(|x| x.is_added(self.ticks.last_change_tick, self.ticks.change_tick))
+    }
+
+    #[inline]
+    fn is_changed(&self) -> bool {
+        self.ticks
+            .component_ticks
+            .as_array()
+            .iter()
+            .any(|x| x.is_changed(self.ticks.last_change_tick, self.ticks.change_tick))
+    }
+
+    #[inline]
+    fn set_changed(&mut self) {
+        for ticks in self.ticks.component_ticks.as_array_mut().iter_mut() {
+            ticks.set_changed(self.ticks.change_tick);
+        }
+    }
+
+    #[inline]
+    fn last_changed(&self) -> u32 {
+        self.ticks.last_change_tick
+    }
+
+    type Inner = AlignedBatchT<T, N, ALIGN>;
+
+    fn set_last_changed(&mut self, last_change_tick: u32) {
+        self.ticks.last_change_tick = last_change_tick;
+    }
+
+    fn bypass_change_detection(&mut self) -> &mut Self::Inner {
+        self.value
+    }
+}
+
+impl<'a, T, const N: usize, const ALIGN: usize> Deref for MutBatch<'a, T, N, ALIGN>
+where
+    ComponentTicks: AlignedBatchGat<N, ALIGN>,
+    T: AlignedBatchGat<N, ALIGN>,
+{
+    type Target = AlignedBatchT<T, N, ALIGN>;
+
+    #[inline]
+    fn deref(&self) -> &Self::Target {
+        self.value
+    }
+}
+
+impl<'a, T, const N: usize, const ALIGN: usize> DerefMut for MutBatch<'a, T, N, ALIGN>
+where
+    ComponentTicks: AlignedBatchGat<N, ALIGN>,
+    T: AlignedBatchGat<N, ALIGN>,
+{
+    #[inline]
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        self.set_changed();
+        self.value
+    }
+}
+
+impl<'a, T, const N: usize, const ALIGN: usize> AsRef<AlignedBatchT<T, N, ALIGN>>
+    for MutBatch<'a, T, N, ALIGN>
+where
+    ComponentTicks: AlignedBatchGat<N, ALIGN>,
+    T: AlignedBatchGat<N, ALIGN>,
+{
+    #[inline]
+    fn as_ref(&self) -> &AlignedBatchT<T, N, ALIGN> {
+        self.deref()
+    }
+}
+
+impl<'a, T, const N: usize, const ALIGN: usize> AsMut<AlignedBatchT<T, N, ALIGN>>
+    for MutBatch<'a, T, N, ALIGN>
+where
+    ComponentTicks: AlignedBatchGat<N, ALIGN>,
+    T: AlignedBatchGat<N, ALIGN>,
+{
+    #[inline]
+    fn as_mut(&mut self) -> &mut AlignedBatchT<T, N, ALIGN> {
+        self.deref_mut()
+    }
+}
+
+//change_detection_impl!(MutBatch<'a, T>, T,);
```

Comment:
Ah; we impl these below. Just remove the commented code.

Reply:
Leave a comment about why we can't use the macro though :)

Reply:
Done!